qa/common/infer_util.py

```diff
@@ -456,3 +456,216 @@ def infer_exact(tester, pf, tensor_shape, batch_size,
+            else:
+                inputs.append(grpcclient.InferInput(
+                    input_name, input_data.shape, np_to_triton_dtype(tensor_dtype)))
+
+            if not (use_cuda_shared_memory or use_system_shared_memory):
+                if config[1] == "http":
+                    inputs[-1].set_data_from_numpy(input_data, binary_data=config[3])
+                else:
+                    inputs[-1].set_data_from_numpy(input_data)
+
+        output_req = []
+        for output_name in expected_dict:
+            if config[1] == "http":
+                output_req.append(httpclient.InferRequestedOutput(
+                    output_name, binary_data=config[3]))
+            else:
+                output_req.append(
+                    grpcclient.InferRequestedOutput(output_name))
+
+        if config[2]:
+            # TODO fix for streaming case
+            continue
+            # results = triton_client.async_stream_infer(model_name,
+            #                                  inputs,
+            #                                  model_version=model_version,
+            #                                  stream=stream,
+            #                                  outputs=output_req)
+        else:
+            results = triton_client.infer(model_name,
+                                          inputs,
+                                          model_version=model_version,
+                                          outputs=output_req,
+                                          request_id=str(_unique_request_id()),
+                                          priority=priority, timeout=timeout_us)
+
+        last_response = results.get_response()
+        if config[1] == "http":
+            if 'error' in last_response:
+                raise InferenceServerException(msg=last_response['error'])
+
+        if config[1] == "http":
+            response_model_name = last_response["model_name"]
+        else:
+            response_model_name = last_response.model_name
+        tester.assertEqual(response_model_name, model_name)
+
+        if model_version != "":
+            if config[1] == "http":
+                response_model_version = last_response["model_version"]
+            else:
+                response_model_version = last_response.model_version
+            tester.assertEqual(response_model_version, model_version)
+
+        if config[1] == "http":
+            response_outputs = last_response["outputs"]
+        else:
+            response_outputs = last_response.outputs
+        tester.assertEqual(len(response_outputs), io_cnt)
+
+        for idx, result in enumerate(response_outputs):
+            if config[1] == "http":
+                result_name = result["name"]
+            else:
+                result_name = result.name
+
+            tester.assertTrue(result_name in expected_dict)
+            if use_system_shared_memory or use_cuda_shared_memory:
+                shm_handle = shm_op_handles[idx]
```

Comment:
Not safe to assume output order in response.

Reply:
We do not run the shared memory variants of the infer_zero tests. This is probably because we cannot create shared memory regions of size 0.

Reply:
Then you can strip the shared memory handling from this. Or you will create a non-empty shared memory region and use it for zero-element tensor. I will suggest to do the latter since it is a plausible use case.

Reply:
Ok, will do that